### PR TITLE
feat: bi-directional streaming map

### PR DESCRIPTION
**Pull request by kohlisid**
Kindly explain what this PR does.

### CODE REVIEW

File: pynumaflow/mapper/_servicer/_async_servicer.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        The pascal case function name comes from the proto map_pb2_grpc.py file.
+        """
+        # proto repeated field(keys) is of type google._upb._message.RepeatedScalarContainer
+        # we need to explicitly convert it to list
+        try:
+            # The first message to be received should be a valid handshake
+            req = await request_iterator.__anext__()
+            # check if it is a valid handshake req
+            if not (req.handshake and req.handshake.sot):
+                raise Exception("MapFn: expected handshake message")
```
Comment (ab93) [inline pynumaflow/mapper/_servicer/_async_servicer.py:44]: can we raise a more specific exception here?

File: pynumaflow/mapper/_servicer/_async_servicer.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+                self._process_inputs(request_iterator, global_result_queue)
+            )
+
+            # keep reading on result queue and send messages back
+            consumer = global_result_queue.read_iterator()
+            async for msg in consumer:
+                # If the message is an exception, we raise the exception
+                if isinstance(msg, BaseException):
+                    await handle_async_error(context, msg)
+                    return
```
Comment (ab93) [inline pynumaflow/mapper/_servicer/_async_servicer.py:60]: We are not really returning anything right? But it should ideally be `AsyncIterable[map_pb2.MapResponse]` correct?

File: pynumaflow/mapper/_servicer/_sync_servicer.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        """
+        Applies a function to each datum element.
+        The pascal case function name comes from the proto map_pb2_grpc.py file.
+        """
+        try:
+            # The first message to be received should be a valid handshake
+            req = next(request_iterator)
+            # check if it is a valid handshake req
+            if not (req.handshake and req.handshake.sot):
+                raise Exception("MapFn: expected handshake message")
```
Comment (ab93) [inline pynumaflow/mapper/_servicer/_sync_servicer.py:43]: same as before

File: pynumaflow/mapper/_servicer/_sync_servicer.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        result_queue: SyncIterator,
+    ):
+        try:
+            d = Datum(
+                keys=list(request.request.keys),
+                value=request.request.value,
+                event_time=request.request.event_time.ToDatetime(),
+                watermark=request.request.watermark.ToDatetime(),
+                headers=dict(request.request.headers),
+            )
```
Comment (ab93) [inline pynumaflow/mapper/_servicer/_sync_servicer.py:107]: Do we need to have this part inside the try except?

File: pynumaflow/mapper/_servicer/_async_servicer.py (hunk 1 above)
Comment (kohlisid) [reply pynumaflow/mapper/_servicer/_async_servicer.py:44]: Sure! Can use a custom error type here

File: pynumaflow/mapper/_servicer/_async_servicer.py (hunk 2 above)
Comment (kohlisid) [reply pynumaflow/mapper/_servicer/_async_servicer.py:60]: Correct, but the error handler would have sent a context error back, hence empty return for exit here.

File: pynumaflow/mapper/_servicer/_sync_servicer.py (hunk 3 above)
Comment (kohlisid) [reply pynumaflow/mapper/_servicer/_sync_servicer.py:43]: Done!

File: pynumaflow/mapper/_servicer/_sync_servicer.py (hunk 4 above)
Comment (kohlisid) [reply pynumaflow/mapper/_servicer/_sync_servicer.py:107]: If there is a type issue, good to catch that mostly

**Review by vigith — APPROVED** [review]

**Pull request merged**